sql/expression/function/date_format.go

```diff
@@ -0,0 +1,316 @@
+}
+
+func dayName(t time.Time) string {
+	return t.Weekday().String()
+}
+
+func yearTwoDigit(t time.Time) string {
+	return strconv.FormatInt(int64(t.Year()) % 100, 10)
+}
+
+type AppendFuncWrapper struct {
+	fn func(time.Time) string
+}
+
+func wrap(fn func(time.Time) string) strftime.Appender {
+	return AppendFuncWrapper{fn}
+}
+
+func (af AppendFuncWrapper) Append(bytes []byte, t time.Time) []byte {
+	s := af.fn(t)
+	return append(bytes, []byte(s)...)
+}
+
+var mysqlDateFormatSpec = strftime.NewSpecificationSet()
+var specifierToFunc = map[byte]func(time.Time) string {
+	'a': nil,
+	'b': nil,
+	'c': monthNum,
+	'D': dayWithSuffix,
+	'd': nil,
+	'e': dayOfMonth,
+	'f': microsecondsStr,
+	'H': nil,
+	'h': twelveHourPadded,
+	'I': twelveHourPadded,
+	'i': minutesStr,
+	'j': nil,
+	'k': twentyFourHourNoPadding,
+	'l': twelveHourNoPadding,
+	'M': fullMonthName,
+	'm': nil,
+	'p': nil,
+	'r': ampmClockStr,
+	'S': nil,
+	's': secondsStr,
+	'T': nil,
+	'U': weekMode0,
+	'u': weekMode1,
+	'V': weekMode2,
+	'v': weekMode3,
+	'W': dayName,
+	'w': nil,
+	'X': yearMode0,
+	'x': yearMode1,
+	'Y': nil,
+	'y': yearTwoDigit,
+}
+
+func init() {
+	for specifier, fn := range specifierToFunc {
+		if fn != nil {
+			panicIfErr(mysqlDateFormatSpec.Set(specifier, wrap(fn)))
+		}
+	}
+
+	// replace any strftime specifiers that aren't supported
+	fn := func(b byte) {
+		if _, ok := specifierToFunc[b]; !ok {
+			panicIfErr(mysqlDateFormatSpec.Set(b, wrap(func(time.Time) string {
+				return string(b)
+			})))
+		}
+	}
+
+	capToLower := byte('a' - 'A')
+	for i := byte('A'); i <= 'Z'; i++ {
+		fn(i)
+		fn(i + capToLower)
+	}
+}
+
+func formatDate(format string, t time.Time) (string, error) {
+	formatter, err := strftime.New(format, strftime.WithSpecificationSet(mysqlDateFormatSpec))
+
+	if err != nil {
+		return "", err
+	}
+
+	return formatter.FormatString(t), nil
+}
+
+// DateFormat function returns a string representation of the date specified in the format specified
+type DateFormat struct {
+	expression.BinaryExpression
+}
+
+// NewDateFormat returns a new DateFormat UDF
+func NewDateFormat(ex, value sql.Expression) sql.Expression {
+	return &DateFormat{
+		expression.BinaryExpression{
+			Left:  ex,
+			Right: value,
+		},
+	}
+}
+
+// Eval implements the Expression interface.
+func (f *DateFormat) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	left, err := f.Left.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if left != nil {
+		return left, nil
+	}
+
+	timeVal, err := sql.Datetime.Convert(left)
+
+	if err != nil {
+		return nil, err
+	}
+
+	t := timeVal.(time.Time)
+
+	right, err := f.Right.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	formatStr, ok := right.(string)
+
+	if !ok {
+		return nil, ErrInvalidArgument.New("DATE_FORMAT", "format must be a string")
+	}
+
+
+	return strftime.Format(formatStr, t)
+}
+
+// Type implements the Expression interface.
+func (f *DateFormat) Type() sql.Type {
```

Comment:
Isn't this just String?

Reply:
yes! Some copy pasta that I forgot to change.